auth_saml/controllers/main.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (C) 2010-2016 XCG Consulting <http://odoo.consulting>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import functools
+import logging
+
+import json as simplejson
+import werkzeug.utils
+
+import odoo
+from odoo import api, _, http, SUPERUSER_ID
+from odoo.http import request
+from odoo import registry as registry_get
+from odoo.addons.web.controllers.main import set_cookie_and_redirect
+from odoo.addons.web.controllers.main import ensure_db
+from odoo.addons.web.controllers.main import login_and_redirect
+from odoo.addons.web.controllers.main import Home
+
+
+_logger = logging.getLogger(__name__)
+
+
+# ----------------------------------------------------------
+# helpers
+# ----------------------------------------------------------
+
+
+def fragment_to_query_string(func):
+    @functools.wraps(func)
+    def wrapper(self, req, **kw):
+        if not kw:
+            return """<html><head><script>
+                var l = window.location;
+                var q = l.hash.substring(1);
+                var r = '/' + l.search;
+                if(q.length !== 0) {
+                    var s = l.search ? (l.search === '?' ? '' : '&') : '?';
+                    r = l.pathname + l.search + s + q;
+                }
+                window.location = r;
+            </script></head><body></body></html>"""
+        return func(self, req, **kw)
+    return wrapper
+
+
+# ----------------------------------------------------------
+# Controller
+# ----------------------------------------------------------
+
+
+class SAMLLogin(Home):
+
+    def list_providers(self):
+        try:
+            providers = request.env['auth.saml.provider'].sudo().search_read(
+                [('enabled', '=', True)])
+        except Exception as e:
+            _logger.exception("SAML2: %s" % str(e))
+            providers = []
+
+        return providers
+
+    @http.route()
+    def web_login(self, *args, **kw):
+        ensure_db()
+        if (
+            request.httprequest.method == 'GET' and
+            request.session.uid and
+            request.params.get('redirect')
+        ):
+
+            # Redirect if already logged in and redirect param is present
+            return http.redirect_with_hash(request.params.get('redirect'))
+
+        providers = self.list_providers()
+
+        response = super(SAMLLogin, self).web_login(*args, **kw)
+        if response.is_qweb:
+            error = request.params.get('saml_error')
+            if error == '1':
+                error = _("Sign up is not allowed on this database.")
+            elif error == '2':
+                error = _("Access Denied")
+            elif error == '3':
```

Comment:
Why not use words to distiguish errors? It'll be easier to trace:

```suggestion
            if error == 'no-signup':
                error = _("Sign up is not allowed on this database.")
            elif error == 'access-denied':
                error = _("Access Denied")
            elif error == 'expired':
```